runtime/criusupport/criusupport.cpp

```diff
@@ -376,16 +392,41 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
 			goto wakeJavaThreads;
 		}
 
+		Trc_CRIU_prepare_internal_hooks(currentThread);
+		vmFuncs->acquireExclusiveVMAccess(currentThread);
+		vm->memoryManagerFunctions->j9mm_iterate_all_objects(vm, PORTLIB, 0, objectIteratorCallback, vm->checkpointState.hookRecords);
+
+		/* Run CRIU hooks before checkpoint - isRestore is FALSE */
+		if (FALSE == vmFuncs->runInternalJVMCriuHooks(vm, FALSE)) {
+			currentExceptionClass = vm->criuSystemCheckpointExceptionClass;
+			nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
+				J9NLS_JCL_CRIU_FAILED_TO_RUN_INTERNAL_CHECKPOINT_HOOKS, NULL);
+			goto wakeJavaThreadsWithExclusiveVMAccess;
+		}
+		Trc_CRIU_before_checkpoint(currentThread);
+
 		systemReturnCode = criu_dump();
+		Trc_CRIU_after_checkpoint(currentThread);
 		if (systemReturnCode < 0) {
 			currentExceptionClass = vm->criuSystemCheckpointExceptionClass;
 			nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_CRIU_DUMP_FAILED, NULL);
-			goto wakeJavaThreads;
+			goto wakeJavaThreadsWithExclusiveVMAccess;
 		}
 
 		/* We can only end up here if the CRIU restore was successful */
 		isAfterCheckpoint = TRUE;
 
+		/* Run CRIU hooks for restore - isRestore is TRUE */
+		if (FALSE == vmFuncs->runInternalJVMCriuHooks(vm, TRUE)) {
```

Comment:
`runInternalJVMCriuHooks(..., TRUE)` and `cleanupCriuHooks` can be in a single call in criuhelpers

Reply:
Moved `cleanupCriuHooks()` into `runInternalJVMCriuHooks(, TRUE)` case. Currently both checkpoint/restore hooks are invoked via same function `runInternalJVMCriuHooks()`, pls advise if two methods are preferred instead.

Reply:
yea, I think two methods is better, one with checkpoint in the name and the other with restore